src/main.rs

```diff
@@ -32,128 +39,133 @@ fn main() {
+                            output_name,
+                            err
+                        );
+                    }
                 }
-                Err(err) => {
-                    log::warn!(
-                        "Skipping '{}' as it might be disconnected: {}",
-                        output_name,
-                        err
-                    );
-
-                    None
-                }
-            }
-        })
-        .collect_vec();
-
-    std::thread::Builder::new()
-        .name("als".to_string())
-        .spawn(move || {
-            let als: Box<dyn als::Als> = match config.als {
-                config::Als::Iio { path, thresholds } => Box::new(
-                    als::iio::Als::new(&path, thresholds)
-                        .expect("Unable to initialize ALS IIO sensor"),
-                ),
-                config::Als::Time { thresholds } => Box::new(als::time::Als::new(thresholds)),
-                config::Als::Webcam { video, thresholds } => Box::new({
-                    let (webcam_tx, webcam_rx) = mpsc::channel();
-                    std::thread::Builder::new()
-                        .name("als-webcam".to_string())
-                        .spawn(move || {
-                            als::webcam::Webcam::new(webcam_tx, video).run();
-                        })
-                        .expect("Unable to start thread: als-webcam");
-                    als::webcam::Als::new(webcam_rx, thresholds)
-                }),
-                config::Als::None { .. } => Box::<als::none::Als>::default(),
-            };
-
-            als::controller::Controller::new(als, als_txs).run();
-        })
-        .expect("Unable to start thread: als");
+
+                (tasks, als_txs)
+            },
+        )
+        .await;
+
+    tasks.push(smol::spawn(async {
+        let mut webcam_task = None;
+        let als: als::Als = match config.als {
+            config::Als::Iio { path, thresholds } => als::Als::Iio(
+                als::iio::Als::new(&path, thresholds)
+                    .await
+                    .expect("Unable to initialize ALS IIO sensor"),
+            ),
+            config::Als::Time { thresholds } => als::Als::Time(als::time::Als::new(thresholds)),
+            config::Als::Webcam { video, thresholds } => als::Als::Webcam({
+                let (webcam_tx, webcam_rx) = channel::bounded(128);
+
+                // TODO: make async
+                webcam_task = Some(smol::unblock(move || {
+                    als::webcam::Webcam::new(webcam_tx, video).run();
+                }));
+                als::webcam::Als::new(webcam_rx, thresholds)
+            }),
+            config::Als::None => als::Als::None(Default::default()),
+        };
+
+        let mut controller = als::controller::Controller::new(als, als_txs);
+
+        smol::future::zip(
+            controller.run(),
+            webcam_task.unwrap_or_else(|| smol::spawn(async {})),
```

Comment:
I'm not too much of a fan of spawning an empty task here just to avoid an if. I think this is a little more confusing than the previous approach, and we'd also be causing some extra runtime load as well as memory overhead because we'd keep the Zip future around. Not to say that this is significant at all, but I just don't think we're gaining anything here.

Reply:
Thank you for challenging this! What do you think of another approach I just pushed, which is basically just to continue pushing only necessary tasks directly to the `tasks` vector? It seems perhaps even simpler than mine and yours, but could I be overlooking some downsides to this idea?